Updates rollback to SDN procedure to add live migration (#77067)
openshift/openshift-docs: modules/nw-ovn-kubernetes-rollback-live.adoc

// Module included in the following assemblies:
//
// * networking/ovn_kubernetes_network_provider/migrate-from-openshift-sdn.adoc

> // * networking/openshift_sdn/rollback-to-ovn-kubernetes.adoc

Reviewer: `rollback-openshift-sdn.adoc`?

:_mod-docs-content-type: PROCEDURE
[id="nw-ovn-kubernetes-rollback-live_{context}"]

> = Using the live migration method to migrate to the OVN-Kubernetes network plugin

Reviewer: `the the Openshift SDN... `?

As a cluster administrator, you can rollback to the OpenShift SDN Container Network Interface (CNI) network plugin by using the live migration method. During the migration with this method, nodes are automatically rebooted and service to the cluster is not interrupted.

[IMPORTANT]
====
You must wait until the migration process from OpenShift SDN to OVN-Kubernetes network plugin is successful before initiating a rollback.
====

> 

Reviewer: Maybe another Important info for the CU might be to inform the user that they can only migrate in 4.16 and no later version of OCP because OpenShift SDN will be depreciated. This will future proof your docs from 4.17 changes to reflect this reality.

Author: @martinkennelly @pliurh Does this mean that we'll only want *this specific procedure* in version 4.16? and we do *not* want it to appear in version 4.17+? That will change the target branch of this pull request.

Reviewer 2: Yes. Live migration will only be available in version 4.16. OpenShiftSDN will be removed in version 4.17. Therefore, users cannot upgrade an SDN cluster to 4.17 without migrating to OVN beforehand. Both Live and offline migration chapters shall be removed from 4.17+ documents.

If a rollback to OpenShift SDN is required, the following table describes the process.

.Performing a rollback to OpenShift SDN
[cols="1,1a",options="header"]
|===

> |User-initiated steps|Migration activity

Reviewer 2: The live migration rollback is similar to the migration. The only different is the target `networkType`.
`oc patch Network.config.openshift.io cluster --type='merge' --patch '{"metadata":{"annotations":{"network.openshift.io/network-type-migration":""}},"spec":{"networkType":"OpenShiftSDN"}}'`

> |Suspend the MCO to ensure that it does not interrupt the migration.

Reviewer: @pliurh do we need to do this step?

// Module included in the following assemblies:
//
// * networking/ovn_kubernetes_network_provider/rollback-to-openshift-sdn.adoc

:_mod-docs-content-type: PROCEDURE
[id="nw-ovn-kubernetes-rollback-live_{context}"]
= Using the live migration method to roll back to the OpenShift SDN network plugin

> As a cluster administrator, you can rollback to the OpenShift SDN Container Network Interface (CNI) network plugin by using the live migration method. During the migration with this method, nodes are automatically rebooted and service to the cluster is not interrupted.

Reviewer 3: ```suggestion
As a cluster administrator, you can roll back to the OpenShift SDN Container Network Interface (CNI) network plugin by using the live migration method. During the rollback with this method, nodes are automatically rebooted and service to the cluster is not interrupted.
```

Author: I change all of these (there were four). I think it was always "rollback" previously and I just copied it over. Fixed now though

[IMPORTANT]
====
You must wait until the migration process from OpenShift SDN to OVN-Kubernetes network plugin is successful before initiating a rollback.
====

If a rollback to OpenShift SDN is required, the following table describes the process.

.Performing a rollback to OpenShift SDN
[cols="1,1a",options="header"]
|===

|User-initiated steps|Migration activity
ifdef::openshift-rosa,openshift-dedicated[]
| Add the `unsupported-red-hat-internal-testing` annotation to the cluster-level network configuration. 
| The Cluster Network Operator (CNO) acknowledges the unsupported testing environment.
endif::[]

> | Patch the cluster-level networking configuration by changing the `networkType` from `OVNKubernetes` to `OpenShiftSDN`.

Reviewer 3: I feel like this maybe should also be a bullet?

| 

> Cluster Network Operator (CNO)::

Reviewer 3: The block like this in the preceding method has an introduction phrase to the list
"Performs the following actions:"

Author: Good catch ty. I had just re-added this section today.

+
--

> * Sets migration related fields in the `network.operator` custom resource (CR) and waits for routable MTUs to be applied to all nodes by MCO.

Reviewer 3: ```suggestion
* Sets migration-related fields in the `network.operator` custom resource (CR) and waits for routable MTUs to be applied to all nodes by Machine Config Operator (MCO).
```

> * Patches the `network.operator` CR to set the migration mode to `Live` for OpenShiftSDN and deploys the OpenShiftSDN network plugin in migration mode.

Reviewer 3: Previous instances are:
```suggestion
* Patches the `network.operator` CR to set the migration mode to `Live` for OpenShift SDN and deploys the OpenShift SDN network plugin in migration mode.
```
As an aside, should "OpenShift SDN" have an attribute to make it easier to keep up with and avoid having "OpenShift" in plain text? If you think it's worthwhile, maybe a thing to do later :)

Author: It's deprecated now and is being removed next release, otherwise I would :)

* Deploys the OpenShift SDN control plane pods.
* Updates the Multus objects to reflect the new network plugin.
--

|
Nodes are automatically rebooted.
|
Cluster:: As nodes reboot, the cluster assigns IP addresses to pods on the OpenShift-SDN network.

|
Enable the MCO after all nodes in the cluster reboot.
|
MCO:: Rolls out an update to the systemd configuration necessary for OpenShift SDN; the MCO updates a single machine per pool at a time by default, so the total time the migration takes increases with the size of the cluster.

> |===

Reviewer 2: This is all for offline migration, instead of live migration.

Author: @pliurh do we want a section similar to this specifically for live migration?

Reviewer 2: Yes. It should be similar to the paragraph in the live migration doc.

```
| The Cluster Network Operator (CNO) acknowledges the unsupported testing environment.
endif::[]

| Patch the cluster-level networking configuration by changing the `networkType` from `OVNKubernetes` to `OpenShiftSDN`.
| 
Cluster Network Operator (CNO)::
+
--
* Sets migration-related fields in the `network.operator` custom resource (CR) and waits for routable MTUs to be applied to all nodes by MCO.
* Patches the `network.operator` CR to set the migration mode to `Live` for OpenShiftSDN and deploys the OpenShiftSDN network plugin in migration mode.
* Deploys OVN-Kubernetes with hybrid overlay enabled.
* Waits for both CNI plugins deployed and updates the conditions in the status of the `network.config` CR.
* Triggers the Machine Config Operator (MCO) to apply the new machine config to each machine config pool, which includes node cordoning, draining, and rebooting.
* Removes migration-related fields from the network. operator CR and performs cleanup actions, such as deleting OpenShift SDN resources and redeploying OVN-Kubernetes in normal mode with the necessary configurations.
* Waits for the OpenShiftSDN redeployment and updates the status conditions in the `network.config` CR to indicate migration completion. 
```

> * Triggers the Machine Config Operator (MCO) to apply the new machine config to each machine config pool, which includes node cordoning, draining, and rebooting.

Reviewer 3: ```suggestion
* Triggers the MCO to apply the new machine config to each machine config pool, which includes node cordoning, draining, and rebooting.
```

> * Install the OpenShift CLI (`oc`).

Reviewer 3: (if you want)
```suggestion
* Install the {oc-first}.
```

Author: Fixed

* A reboot can be triggered manually for each node.

> * The cluster is in a known good state, without any errors.

Reviewer 3: can these be rephrased such that they are parallel construction?

> .Procedure

Reviewer 3: Generally, it's a good practice to have steps be goal-first (so, "To do XYZ, run the following command" or "Do XYZ by running the following command"). That would also make these more consistent w/ the rest of the docs overall.

. Enter the following command to initiate the rollback to OpenShift SDN:
+
[source,terminal]
----
$ oc patch Network.config.openshift.io cluster --type='merge' --patch '{"metadata":{"annotations":{"network.openshift.io/network-type-migration":""}},"spec":{"networkType":"OpenShiftSDN"}}'
----

. You can enter the following command to watch the progress of your migration:
+
[source,terminal]
----
$ watch -n1 'oc get network.config/cluster -o json | jq ".status.conditions[]|\"\\(.type) \\(.status) \\(.reason) \\(.message)\""  -r | column --table --table-columns NAME,STATUS,REASON,MESSAGE --table-columns-limit 4; echo; oc get mcp -o wide; echo; oc get node -o "custom-columns=NAME:metadata.name,STATE:metadata.annotations.machineconfiguration\\.openshift\\.io/state,DESIRED:metadata.annotations.machineconfiguration\\.openshift\\.io/desiredConfig,CURRENT:metadata.annotations.machineconfiguration\\.openshift\\.io/currentConfig,REASON:metadata.annotations.machineconfiguration\\.openshift\\.io/reason"'
----
+
The command prints the following information every second:
+
* The conditions on the status of the `network.config.openshift.io/cluster` object, reporting the progress of the migration.

> * The status of different nodes with respect to the `machine-config-operator` resource, including whether they are upgrading or have been upgraded, as well as their current and desired configurations.

Reviewer 2: Once the live rollback is done, we shall ask users to remove the annotation `network.openshift.io/network-type-migration=` from the network.config CR.

Author: @pliurh please check

Reviewer 2: The statement is correct. But the command needs to be updated.
> oc annotate network.config cluster network.openshift.io/network-type-migration-

Author: @pliurh updated command.
